libecole/src/scip/scimpl.cpp

```diff
@@ -141,5 +245,69 @@ auto ReverseBranchrule::scip_execlp(SCIP* scip, SCIP_BRANCHRULE* /*branchrule*/,
 	return action_func(scip, result);
 }
 
+}  // namespace
+
+/*******************************
+ *  Definition of ReverseHeur  *
+ *******************************/
+
+namespace {
+
+scip::ReverseHeur::ReverseHeur(
+	SCIP* scip,
+	std::weak_ptr<utility::Controller::Executor> weak_executor_,
+	int trials_per_node_,
+	int depth_freq,
+	int depth_start,
+	int depth_stop) :
+	::scip::ObjHeur(
+		scip,
+		scip::ReverseHeur::name,
+		"Primal heuristic that waits for another thread to provide a primal solution.",
+		'e',
+		scip::ReverseHeur::max_priority,
+		depth_freq,
+		depth_start,
+		depth_stop,
+		SCIP_HEURTIMING_AFTERNODE,
+		false),
+	weak_executor(std::move(weak_executor_)),
+	trials_per_node(trials_per_node_) {}
+
+auto ReverseHeur::scip_exec(
+	SCIP* scip,
+	SCIP_HEUR* /*heur*/,
+	SCIP_HEURTIMING /*heurtiming*/,
+	SCIP_Bool /*nodeinfeasible*/,
+	SCIP_RESULT* result) -> SCIP_RETCODE {
+	*result = SCIP_DIDNOTRUN;
+	auto retcode = SCIP_OKAY;
+
+	for (int trial = 0; trial < trials_per_node || trials_per_node < 0; trial++) {
```

Comment:
I agree it would be nice to have all the logic in one place, but I don't really see how to do it with the way the controller / executer thing works... Let's discuss that in our next meeting.